File: scripts/research/f017_dprefix_exact_scaffold.c
#include <errno.h>
#include <fcntl.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>

/* DPREFIX-EXACT-1 implementation A.
 *
 * Every reduction is a single, increasing-index binary64 accumulator.  The
 * reviewed build disables contraction and vectorization.  Position-zero
 * attention has exactly one visible token, so q/k cannot affect the value
 * path and are analytically eliminated; all value-path and FFN contractions
 * remain explicit.
 */

enum { HIDDEN = 6144, HEADS = 64, VALUE = 256, KV = 512, FFN = 12288 };
static const double RMS_EPSILON = 9.999999747378752e-6;

typedef struct {
    char name[96];
    char path[1024];
    size_t dims[3];
    int rank;
    size_t bytes;
    const float *data;
} Tensor;

static Tensor tensors[40];
static size_t tensor_count = 0;

static void fail(const char *message) {
    fprintf(stderr, "DPREFIX-EXACT-1-A: %s\n", message);
    exit(2);
}

static Tensor *tensor(const char *name) {
    for (size_t i = 0; i < tensor_count; ++i) {
        if (strcmp(tensors[i].name, name) == 0) return &tensors[i];
    }
    fail("missing tensor");
    return NULL;
}

static void load_manifest(const char *path) {
    FILE *source = fopen(path, "r");
    if (!source) fail("manifest open");
    char line[1400];
    while (fgets(line, sizeof line, source)) {
        if (tensor_count == 40) fail("manifest overflow");
        Tensor *item = &tensors[tensor_count];
        unsigned long long d0 = 0, d1 = 0, d2 = 0;
        int fields = sscanf(line, "%95s\t%1023s\t%d\t%llu\t%llu\t%llu",
                            item->name, item->path, &item->rank, &d0, &d1, &d2);
        if (fields != 6 || item->rank < 1 || item->rank > 3) fail("manifest parse");
        item->dims[0] = (size_t)d0; item->dims[1] = (size_t)d1; item->dims[2] = (size_t)d2;
        size_t count = item->dims[0];
        for (int i = 1; i < item->rank; ++i) count *= item->dims[i];
        item->bytes = count * sizeof(float);
        int fd = open(item->path, O_RDONLY | O_NOFOLLOW);
        if (fd < 0) fail("tensor open");
        struct stat status;
        if (fstat(fd, &status) || !S_ISREG(status.st_mode) || (size_t)status.st_size != item->bytes)
            fail("tensor identity");
        void *mapped = mmap(NULL, item->bytes, PROT_READ, MAP_PRIVATE, fd, 0);
        close(fd);
        if (mapped == MAP_FAILED) fail("tensor mmap");
        item->data = (const float *)mapped;
        ++tensor_count;
    }
    fclose(source);
    if (tensor_count != 40) fail("manifest count");
}

static float *allocate(size_t count) {
    float *value = calloc(count, sizeof(float));
    if (!value) fail("allocation");
    return value;
}

static void matvec(const float *matrix, size_t rows, size_t columns,
                   const float *vector, float *output) {
    for (size_t row = 0; row < rows; ++row) {
        double sum = 0.0;
        const float *weights = matrix + row * columns;
        for (size_t column = 0; column < columns; ++column)
            sum += (double)weights[column] * (double)vector[column];
        output[row] = (float)sum;
    }
}

static void rms_norm(const float *input, const float *weight, size_t count, float *output) {
    double square_sum = 0.0;
    for (size_t i = 0; i < count; ++i) square_sum += (double)input[i] * (double)input[i];
    float inverse = (float)(1.0 / sqrt(square_sum / (double)count + RMS_EPSILON));
    for (size_t i = 0; i < count; ++i) output[i] = (input[i] * inverse) * weight[i];
}

static void write_surface(const char *directory, const char *name, const float *values, size_t count) {
    char path[1200];
    snprintf(path, sizeof path, "%s/%s.f32le", directory, name);
    FILE *target = fopen(path, "wb");
    if (!target || fwrite(values, sizeof(float), count, target) != count || fclose(target))
        fail("surface write");
}

static float *run_layer(int layer, const float *residual, const char *output_dir) {
    char name[96];
    float *normalized = allocate(HIDDEN);
    snprintf(name, sizeof name, "blk.%d.attn_norm.weight", layer);
    rms_norm(residual, tensor(name)->data, HIDDEN, normalized);

    float *kv_raw = allocate(KV + 64);
    snprintf(name, sizeof name, "blk.%d.attn_kv_a_mqa.weight", layer);
    matvec(tensor(name)->data, KV + 64, HIDDEN, normalized, kv_raw);
    free(normalized);
    float *kv = allocate(KV);
    snprintf(name, sizeof name, "blk.%d.attn_kv_a_norm.weight", layer);
    rms_norm(kv_raw, tensor(name)->data, KV, kv);
    free(kv_raw);

    float *values = allocate(HEADS * VALUE);
    snprintf(name, sizeof name, "blk.%d.attn_v_b.weight", layer);
    const float *value_weights = tensor(name)->data;
    for (size_t head = 0; head < HEADS; ++head)
        matvec(value_weights + head * VALUE * KV, VALUE, KV, kv, values + head * VALUE);
    free(kv);

    float *attention = allocate(HIDDEN);
    snprintf(name, sizeof name, "blk.%d.attn_output.weight", layer);
    matvec(tensor(name)->data, HIDDEN, HEADS * VALUE, values, attention);
    free(values);
    char surface[64];
    snprintf(surface, sizeof surface, "layer_%d_attention", layer);
    write_surface(output_dir, surface, attention, HIDDEN);

    float *attention_residual = allocate(HIDDEN);
    for (size_t i = 0; i < HIDDEN; ++i) attention_residual[i] = residual[i] + attention[i];
    free(attention);
    float *ffn_input = allocate(HIDDEN);
    snprintf(name, sizeof name, "blk.%d.ffn_norm.weight", layer);
    rms_norm(attention_residual, tensor(name)->data, HIDDEN, ffn_input);

    float *gate = allocate(FFN), *up = allocate(FFN), *activated = allocate(FFN);
    snprintf(name, sizeof name, "blk.%d.ffn_gate.weight", layer);
    matvec(tensor(name)->data, FFN, HIDDEN, ffn_input, gate);
    snprintf(name, sizeof name, "blk.%d.ffn_up.weight", layer);
    matvec(tensor(name)->data, FFN, HIDDEN, ffn_input, up);
    free(ffn_input);
    for (size_t i = 0; i < FFN; ++i)
        activated[i] = (gate[i] / (1.0f + expf(-gate[i]))) * up[i];
    free(gate); free(up);
    float *down = allocate(HIDDEN);
    snprintf(name, sizeof name, "blk.%d.ffn_down.weight", layer);
    matvec(tensor(name)->data, HIDDEN, FFN, activated, down);
    free(activated);
    float *output = allocate(HIDDEN);
    for (size_t i = 0; i < HIDDEN; ++i) output[i] = attention_residual[i] + down[i];
    free(attention_residual); free(down);
    snprintf(surface, sizeof surface, "layer_%d_output", layer);
    write_surface(output_dir, surface, output, HIDDEN);
    return output;
}

int main(int argc, char **argv) {
    if (argc != 3) fail("usage: exact-input-manifest output-directory");
    load_manifest(argv[1]);
    Tensor *embedding = tensor("token_embd.weight");
    if (embedding->rank != 2 || embedding->dims[0] != 154880 || embedding->dims[1] != HIDDEN)
        fail("embedding shape");
    float *hidden = allocate(HIDDEN);
    memcpy(hidden, embedding->data + (size_t)9703 * HIDDEN, HIDDEN * sizeof(float));
    write_surface(argv[2], "embedding", hidden, HIDDEN);
    for (int layer = 0; layer < 3; ++layer) {
        float *next = run_layer(layer, hidden, argv[2]);
        free(hidden); hidden = next;
    }
    write_surface(argv[2], "layer_3_entry", hidden, HIDDEN);
    free(hidden);
    return 0;
}

File: scripts/research/f017_dprefix_exact_scaffold.rs
use std::collections::BTreeMap;
use std::convert::TryInto;
use std::fs::{self, File};
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::{Path, PathBuf};

const HIDDEN: usize = 6144;
const HEADS: usize = 64;
const VALUE: usize = 256;
const KV: usize = 512;
const FFN: usize = 12288;
const RMS_EPSILON: f64 = 9.999999747378752e-6;

#[derive(Clone)]
struct Tensor { path: PathBuf, dims: Vec<usize> }

fn die(message: &str) -> ! { eprintln!("DPREFIX-EXACT-1-B: {message}"); std::process::exit(2) }

fn manifest(path: &Path) -> BTreeMap<String, Tensor> {
    let text = fs::read_to_string(path).unwrap_or_else(|_| die("manifest read"));
    let mut result = BTreeMap::new();
    for line in text.lines() {
        let fields: Vec<_> = line.split('\t').collect();
        if fields.len() != 6 { die("manifest parse") }
        let rank: usize = fields[2].parse().unwrap_or_else(|_| die("manifest rank"));
        let all = [fields[3], fields[4], fields[5]].map(|x| x.parse::<usize>().unwrap_or_else(|_| die("manifest dimension")));
        let dims = all[..rank].to_vec();
        let expected = dims.iter().product::<usize>() * 4;
        if fs::metadata(fields[1]).map(|m| m.len() as usize).unwrap_or(0) != expected { die("tensor identity") }
        if result.insert(fields[0].to_string(), Tensor { path: fields[1].into(), dims }).is_some() { die("duplicate tensor") }
    }
    if result.len() != 40 { die("manifest count") }
    result
}

fn load(tensors: &BTreeMap<String, Tensor>, name: &str) -> Vec<f32> {
    let item = tensors.get(name).unwrap_or_else(|| die("missing tensor"));
    let raw = fs::read(&item.path).unwrap_or_else(|_| die("tensor read"));
    raw.chunks_exact(4).map(|b| f32::from_le_bytes(b.try_into().unwrap())).collect()
}

fn embedding(tensors: &BTreeMap<String, Tensor>) -> Vec<f32> {
    let item = tensors.get("token_embd.weight").unwrap_or_else(|| die("embedding absent"));
    if item.dims != [154880, HIDDEN] { die("embedding shape") }
    let mut file = File::open(&item.path).unwrap_or_else(|_| die("embedding open"));
    file.seek(SeekFrom::Start((9703 * HIDDEN * 4) as u64)).unwrap_or_else(|_| die("embedding seek"));
    let mut raw = vec![0u8; HIDDEN * 4];
    file.read_exact(&mut raw).unwrap_or_else(|_| die("embedding read"));
    raw.chunks_exact(4).map(|b| f32::from_le_bytes(b.try_into().unwrap())).collect()
}

fn matvec(matrix: &[f32], rows: usize, columns: usize, vector: &[f32]) -> Vec<f32> {
    if matrix.len() != rows * columns || vector.len() != columns { die("matvec shape") }
    let mut output = vec![0.0f32; rows];
    for row in 0..rows {
        let mut sum = 0.0f64;
        for column in 0..columns { sum += matrix[row * columns + column] as f64 * vector[column] as f64; }
        output[row] = sum as f32;
    }
    output
}

fn rms(input: &[f32], weight: &[f32]) -> Vec<f32> {
    if input.len() != weight.len() { die("rms shape") }
    let mut sum = 0.0f64;
    for &value in input { sum += value as f64 * value as f64; }
    let inverse = (1.0 / (sum / input.len() as f64 + RMS_EPSILON).sqrt()) as f32;
    input.iter().zip(weight).map(|(&x, &w)| (x * inverse) * w).collect()
}

fn write_surface(root: &Path, name: &str, values: &[f32]) {
    let mut file = File::create(root.join(format!("{name}.f32le"))).unwrap_or_else(|_| die("surface create"));
    for value in values { file.write_all(&value.to_le_bytes()).unwrap_or_else(|_| die("surface write")); }
}

fn layer(tensors: &BTreeMap<String, Tensor>, layer: usize, residual: &[f32], output: &Path) -> Vec<f32> {
    let p = format!("blk.{layer}");
    let normalized = rms(residual, &load(tensors, &format!("{p}.attn_norm.weight")));
    let kv_raw = matvec(&load(tensors, &format!("{p}.attn_kv_a_mqa.weight")), KV + 64, HIDDEN, &normalized);
    let kv = rms(&kv_raw[..KV], &load(tensors, &format!("{p}.attn_kv_a_norm.weight")));
    let value_weights = load(tensors, &format!("{p}.attn_v_b.weight"));
    let mut values = vec![0.0f32; HEADS * VALUE];
    for head in 0..HEADS {
        let begin = head * VALUE * KV;
        let result = matvec(&value_weights[begin..begin + VALUE * KV], VALUE, KV, &kv);
        values[head * VALUE..(head + 1) * VALUE].copy_from_slice(&result);
    }
    let attention = matvec(&load(tensors, &format!("{p}.attn_output.weight")), HIDDEN, HEADS * VALUE, &values);
    write_surface(output, &format!("layer_{layer}_attention"), &attention);
    let attention_residual: Vec<f32> = residual.iter().zip(&attention).map(|(&a, &b)| a + b).collect();
    let ffn_input = rms(&attention_residual, &load(tensors, &format!("{p}.ffn_norm.weight")));
    let gate = matvec(&load(tensors, &format!("{p}.ffn_gate.weight")), FFN, HIDDEN, &ffn_input);
    let up = matvec(&load(tensors, &format!("{p}.ffn_up.weight")), FFN, HIDDEN, &ffn_input);
    let activated: Vec<f32> = gate.iter().zip(&up).map(|(&g, &u)| (g / (1.0 + (-g).exp())) * u).collect();
    let down = matvec(&load(tensors, &format!("{p}.ffn_down.weight")), HIDDEN, FFN, &activated);
    let result: Vec<f32> = attention_residual.iter().zip(&down).map(|(&a, &b)| a + b).collect();
    write_surface(output, &format!("layer_{layer}_output"), &result);
    result
}

fn main() {
    let args: Vec<_> = std::env::args_os().collect();
    if args.len() != 3 { die("usage: exact-input-manifest output-directory") }
    let tensors = manifest(Path::new(&args[1]));
    let output = Path::new(&args[2]);
    fs::create_dir_all(output).unwrap_or_else(|_| die("output directory"));
    let mut hidden = embedding(&tensors);
    write_surface(output, "embedding", &hidden);
    for index in 0..3 { hidden = layer(&tensors, index, &hidden, output); }
    write_surface(output, "layer_3_entry", &hidden);
}
